Move globus app (#1076)
globus/globus-sdk-python: docs/authorization/globus_app/apps.rst

.. _globus_apps:

.. currentmodule:: globus_sdk

GlobusApps
==========

:class:`GlobusApp` is a high level construct designed to simplify the process of
of authenticating for Globus service clients.

A ``GlobusApp`` requires a programmatic Globus Auth entity called a "client" to be
supplied as the requester driving authentication flows. A client can be created and

> managed through the `Globus Developer Console <https://developers.globus.org/>`_.

Reviewer: I'm open to debating this, but I think we should send users to

```suggestion
managed through the `Globus Developer Console <https://app.globus.org/settings/developers>`_.
```

now that it's integrated into the webapp.

Author: Yeah, I didn't know which was more desirable; I copied from other locations in the SDK (and have used this link in past PRs) but for the files I've touched as a part of this change I can update to using the app-based one.

Author: I just opened a thread in #auth to confirm what they'd like us to do.

Author: Confirmed with Brigitte that `https://app.globus.org/settings/developers` should replace all instances of `https://developers.globus.org/`.

Updated all of the usages in this PR.

There are two types of GlobusApps; :class:`UserApp`, for contexts in which a client
should submit requests on behalf of a user, and :class:`ClientApp`, for contexts which

> a client should submit requests on behalf of themselves.

Reviewer: Because this is intro text, I'm giving it a bit of extra scrutiny. I think the notion of "clients acting on behalf of themselves" is very abstract and hard for a new user to grasp.

I would phrase these in terms of "login to the app" and "the client acts as an automated user", leading to the following rephrase:
> ```rst
> There are two types of GlobusApp: :class:`UserApp` and :class:`ClientApp`.
> A :class:`UserApp` requires users to login to the application, allowing the application to
> submit requests on behalf of the user.
> A :class:`ClientApp` uses its own credentials to act as a user, and is suitable
> for a variety of automation use cases.
> ```

Author: Rephrased it again...

Reviewer: FWIW, your rewrite is fantastic!

Marking this thread resolved.

Author: Glad to hear you like it! Thanks for pushing me on it.

I'm going to unresolve this thread actually; I pinged Ada & Brigitte about this section to get their sign off as well. Want to leave it open in case they give comments here.

The interfaces of these classes, defined below, intentionally include many
"sane defaults" (i.e., storing oauth2 access tokens in a json file). These defaults

> however can be overridden if the behavior is not desired. For more information on what
> you can customize and how, see :ref:`globus_app_config`.

Reviewer: ```suggestion
can be overridden to customize the app's behavior. For more information on what
you can customize and how, see :ref:`globus_app_config`.
```

.. _globus_apps:

.. currentmodule:: globus_sdk

GlobusApps
==========

.. warning::

>     Currently ``GlobusApp`` only supports scripting-based use cases.

Reviewer 2: Workshopping...

```suggestion
    Currently, ``GlobusApp`` can be used in scripts (e.g., notebooks or automation) and local applications launched directly by a user.
```

Reviewer: If we're making notes on this warning, I would like to see it converted from a `warning` to a `note` to be less of an attention-grab at the start of the doc.

I might phrase it as
> ``GlobusApp`` only supports client applications, where a user or client logs in or authenticates.
> Resource server implementations should not use ``GlobusApp``.

This phrasing reveals my belief that any abstraction aimed at Resource Server implementations will be fully distinct from ``GlobusApp``.

Author: Reclassified as a note & updated the wording here; I pulled from both of your suggestions for the final copy.

>     Hosted applications (i.e. web servers) should see :ref:`globus_authorizers` instead.

Reviewer: I believe that a hosted webapp which uses 3-legged oauth to authenticate users can be developed on top of `GlobusApp` but requires that you implement your own login flow manager.

This is a case we haven't discussed much, but I haven't seen it as necessary to build explicit support, since it will always require custom login flow code.

But I think there's some blurring of lines happening (already in the current doc) here: a hosted client application and a resource server are different cases, and should be described as distinct.

>     Hosted applications (i.e. web servers) should see :ref:`globus_authorizers` instead.

Author: Web services that require dependent tokens effectively don't work with GlobusApp today.

Somebody probably could encapsulate that behavior in a login flow manager, but I would be very surprised if they did (and did it well). Given that we do have very plug-and-play interfaces for scripting cases I'm of the opinion that our messaging is "this works for scripting, not web services".

>     Hosted applications (i.e. web servers) should see :ref:`globus_authorizers` instead.

Reviewer: > Web services that require dependent tokens effectively don't work with GlobusApp today.

That's not what I'm describing as what should work -- that's the resource server case.

I'm talking about a web application which is a pure client, and does the 3-legged OAuth login flow (authorization code grant). Think something like a heavyweight django app which is a client to Globus services.
That's a hosted web app but it's not a resource server.

>     Hosted applications (i.e. web servers) should see :ref:`globus_authorizers` instead.

Reviewer 2: ```suggestion
    Web services (and other hosted applications that act as resource servers) should see :ref:`globus_authorizers` instead.
```

> :class:`GlobusApp` is a high level construct designed to simplify the process of
> of authenticating for Globus service clients.

Reviewer 2: ```suggestion
:class:`GlobusApp` is a high level construct designed to simplify authentication
for Globus service clients.
```

Reviewer: I'm not sure why, but this rephrase is highlighting for me something else:
> Globus service clients.

"service clients" might mislead some folks into thinking that a "service client" is something meaningful vis-a-vis Globus Auth.

Perhaps...

> authentication for applications using globus-sdk clients.

?

Author: I'm hapyp to replace "Globus" with "globus-sdk" but I want to keep "service clients".

The term "client" is overloaded , particularly in this section of the docs so I want to be as clear as possible about what type of client we mean. I'll make the term "service client" be a link to the "[Service Clients](https://globus-sdk-python.readthedocs.io/en/stable/services/index.html)" section of our docs which hopefully will alleviate the concern that it's a Globus Auth thing.

Reviewer: > The term "client" is overloaded

No contest!

:+1: for <code>globus-sdk :ref:&#96;service clients <service_clients>&#96;</code>

I'm not sure if we ever intentionally made "service client" a term of art -- other than it being the heading on that page, I don't think the term is used anywhere. And I would tend to pedantically call them "globus-sdk client classes" or "globus-sdk client objects" in most contexts. But I don't really object to the terminology and -- hey! -- it's the name used in the `services/index.rst` doc.

Author: Ended up going with `"globus-sdk services" where "services" is a link to our "service clients" page.

I also made an effort to qualify other usages of "client" across the page as "OAuth2 client"

> A ``GlobusApp`` will always require a programmatic entity called a "client". Depending
> on the type of ``GlobusApp`` in use, the client will play different roles in the
> request flow.

Reviewer 2: ```suggestion
A ``GlobusApp`` requires a programmatic entity called a "client". Depending
on the type of ``GlobusApp`` in use, the client will play different roles in the
request flow.
```

Reviewer: > Wondering if the first use of client here needs to link out to the relevant Auth docs: https://docs.globus.org/api/auth/developer-guide/#developing-apps

I vote (strongly) against this. We should have SDK-specific documentation on this subject, and having two "sources of truth" for this information will confuse users.

> ...also wondering if we should indicate which of the options from the provided list "Service Account, Portal / Science Gateway, Thick Client, API, or Advanced Application" are relevant?

:+1: for this.
I have some minor gripes about those categories being what they are[^1], but I think matching the web experience in our docs will help users.

[^1]: This part of the webapp is only for developers, but intentionally choosing use-case names rather than names meant to match the SDKs and Auth docs _assumes_ that the user in question is non-technical. It's dissonant and I don't see that it was necessary to introduce this discrepancy. I strongly believe that this is a poor UX decision, but I only minorly care about it.

Reviewer 2: I'm totally okay with not doing the first suggestion if we do the second 😄 on reflection, I think that's actually the one I care more about.

Author: Refactored this section a good amount; the linkage to specific webapp based terminology (point two) is included now in the comparison table.

> *   A :class:`UserApp` drives interaction starting with a browser based login
>     flow. With this type of ``GlobusApp``, newly created resources (collections,
>     flows, etc.) are owned by the human user who has logged in (and thus granted the
>     app authorization tokens). Similarly resource access is evaluated based on that
>     user's permissions.

Reviewer 2: ```suggestion
*   A :class:`UserApp` drives interaction using a web-browser-based login
    flow. With this type of ``GlobusApp``, all service interactions are performed as the
    Globus-identity of the user who logged in. For instance, newly created resources
    (collections, flows, etc.) may be owned by that user. Similarly, access to resources
    will be evaluated based on that user's permissions.
```

Author: This is now in the table with updated wording.

>     A :class:`UserApp` should be used in cases where a script simplifies a process
>     for a user, but the user is ultimately responsible for the actions taken by the
>     script. The `Globus CLI <https://docs.globus.org/cli/>`_ is a good example of
>     one such use case.

Reviewer 2: ```suggestion
    A :class:`UserApp` is appropriate when a script or local application simplifies a process
    for a user, but the user is still ultimately responsible for the actions performed.
    The `Globus CLI <https://docs.globus.org/cli/>`_ is one such example.
```

Author: The content for this piece now reads (in a table):

> Commonly used to perform actions as a specific end user (e.g., the `Globus CLI <https://docs.globus.org/cli/>`_)

Author: I switched back to "appropriate for"; with the other wording shift, I do think that "appropriate" is more accurate and reads better than "commonly used to"

>     A :class:`UserApp` **should use a "native"-type client**. "Confidential"-type
>     clients are allowed, but prove more difficult to configure.

Reviewer 2: ```suggestion
    A :class:`UserApp` **should use a "native" client**. "Confidential"
    clients are allowed, but prove more difficult to configure.
```

Reviewer 2: Follow-up question: Is there _ever_ a compelling reason to use a confidential client in this case? (i.e., if there is a situation where that's actually the correct solution, it _might_ be useful to say something like "Confidential clients are allowed but should only be used if you need to ...")

Reviewer: My short answer here is "yes".

There are only two specific cases I know of: 
- a hosted client (e.g., a Django app)
- a templated client (e.g., actual cli implementation) 

I still wonder a little if we should split UserApp into multiple classes for the different cases, but I figured that we can add subclasses in the future if we want.

Author: This concept is reflected in the table now.

> *   A :class:`ClientApp` by contrast does not require any manual login flow. Instead
>     newly created resources (collections, flows, etc.) are owned by the client
>     itself, not a human. Similarly, resource access will be evaluated based on the
>     client's permissions.

Reviewer 2: ```suggestion
*   A :class:`ClientApp` does not use a login flow.
    All service interactions are performed as the application itself, rather
    than as a Globus user's identity. Resources created (collections, flows, etc.)
    are may be owned by the application's client. Similarly, resource access will be
    evaluated based on the application's client's permissions.
```

>     A :class:`ClientApp` should be used in cases where a service account, potentially
>     managed by multiple humans, is responsible for the actions taken by the script or
>     process. These cases mostly involve some kind of automation, i.e., frequent data
>     movement or flow invocation.

Reviewer 2: ```suggestion
    A :class:`ClientApp` is appropriate when a service account, potentially
    managed by multiple humans, is responsible for actions taken by the script or
    process. These cases mostly involve some kind of automation, e.g., frequent data
    movement or flow invocation.
```

>     A :class:`ClientApp` **must use a "confidential"-type client**. "Native"-type
>     clients don't have the necessary security properties (a client secret) to own their
>     own resources and will be rejected.

Reviewer 2: ```suggestion
    A :class:`ClientApp` **must use a "confidential" client**. "Native"
    clients don't have the necessary security properties (a client secret) to manage
    resources and will be rejected.
```

    .. note::

        Not all Globus services or contexts support ``ClientApp``-based interactions.

>         GCS collections with sensitive data, for instance, may require that a human user
>         has authenticated with their university email recently before allowing data
>         interaction. In these cases, a ``UserApp`` must be used instead.

Reviewer 2: ```suggestion
        GCS collections with sensitive data, for instance, may require that a Globus user
        has authenticated with their university email recently before allowing data
        interaction. In these cases, a ``UserApp`` must be used instead.
```

> Both "native" and "confidential" type clients, may be created and managed as a part
> of a `Globus Project <https://app.globus.org/settings/developers>`_.

Reviewer 2: ```suggestion
Both "native" and "confidential" clients may be created and managed as part
of a `Globus Project <https://app.globus.org/settings/developers>`_.
```

.. _globus_apps:

.. currentmodule:: globus_sdk

GlobusApps
==========

.. note::

    Currently ``GlobusApp`` can only be used in scripts (e.g., notebooks or automation)
    and local applications launched directly by a user.

    Web services and other hosted applications operating as resource servers should see
    :ref:`globus_authorizers` instead.

:class:`GlobusApp` is a high level construct designed to simplify authentication for
interactions with :ref:`globus-sdk services <services>`.

A ``GlobusApp`` uses an OAuth2 client to obtain and manage OAuth2 tokens required for
API interactions. OAuth2 clients must be created external to the SDK by registering an
application at the `Globus Developer's Console <https://app.globus.org/developers>`_.

The following section provides a comparison of of the specific types of ``GlobusApps``
to aid in selecting the proper one for your use case.

Types of GlobusApps
-------------------

There are two types of ``GlobusApps`` available in the SDK: :class:`UserApp` and

> :class:`ClientApp`, each modeling a different style of service interaction:

Author: Pulled this into its own sentence.

.. list-table::
    :widths: 50 50
    :header-rows: 1

    *   - **UserApp**

>         - **ClientApp**

Reviewer: I looked over at the docs prior to this PR, and I'm wondering why we aren't relying on the definitions which are currently available in the `Using a GlobusApp` doc. They read extremely well to me:

> There are two flavors of GlobusApp:
>
> - [UserApp](https://globus-sdk-python.readthedocs.io/en/stable/experimental/globus_app.html#globus_sdk.experimental.globus_app.UserApp), a GlobusApp for interactions between an end user and Globus services. Operations are performed as a user identity.
>
> - [ClientApp](https://globus-sdk-python.readthedocs.io/en/stable/experimental/globus_app.html#globus_sdk.experimental.globus_app.ClientApp), a GlobusApp for interactions between a client (i.e. service account) and Globus services. Operations are performed as a client identity.

These definitions are both accurate and short.
I think we should start with these at the top-level, and that any details on things like mapping use-cases or behaviors to the two types should go somewhere later.

>         - **ClientApp**

Author: Added these definitions to the start.

    *   - Appropriate for simplifying a process otherwise performed by a user (e.g.,

>           the `Globus CLI <https://docs.globus.org/cli/>`_)

Reviewer: My view has refined and softened as I've become more of a proponent of making the definitions separate.

        - Appropriate for automation against non-sensitive data

    *   - OAuth2 tokens are obtained by logging a human in (through a web browser)
        - OAuth2 tokens are obtained by programmatically exchanging an OAuth2 client's
          secret

    *   - Created resources (e.g., collections or flows) by default are owned by a user
        - Created resources (e.g., collections or flows) by default are owned by the
          OAuth2 client

    *   - Existing resource access is evaluated based on a user's permissions
        - Existing resource access is evaluated based on the OAuth2 client's permissions

    *   - Should use a "native" OAuth2 client

          (`Register a thick client

>           <https://app.globus.org/settings/developers/registration/public_installed_client>`_)

Author: Expanded this to `Should typically use a "native"... May use a "confidential"...`
